subprojects/core/src/integTest/groovy/org/gradle/api/provider/ProviderIntegrationTest.groovy

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.provider
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Unroll
+
+import static org.gradle.util.TextUtil.normaliseFileSeparators
+
+class ProviderIntegrationTest extends AbstractIntegrationSpec {
+
+    private static final String OUTPUT_FILE_CONTENT = 'Hello World!'
+    File defaultOutputFile
+    File customOutputFile
+
+    def setup() {
+        defaultOutputFile = file('build/output.txt')
+        customOutputFile = file('build/custom.txt')
+    }
+
+    @Unroll
+    def "can create and use provider by custom task written as #language class"() {
+        given:
+        file("buildSrc/src/main/$language/MyTask.${language}") << taskTypeImpl
+        buildFile << """
+            task myTask(type: MyTask)
+        """
+
+        when:
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+
+        when:
+        buildFile << """
+             myTask {
+                enabled = provider(true)
+                outputFiles = provider(files("${normaliseFileSeparators(customOutputFile.canonicalPath)}"))
+            }
+        """
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+        customOutputFile.isFile()
+        customOutputFile.text == OUTPUT_FILE_CONTENT
+
+        where:
+        language | taskTypeImpl
+        'groovy' | customGroovyBasedTaskType()
+        'java'   | customJavaBasedTaskType()
+    }
+
+    def "can assign default values for provider properties"() {
+        given:
+        buildFile << """
+            task myTask(type: MyTask)
+
+            class MyTask extends DefaultTask {
+                private Provider<Boolean> enabled = project.defaultProvider(Boolean)
+                private Provider<FileCollection> outputFiles = project.defaultProvider(FileCollection)
```

Comment:
Would this approach actually work for mapping extension properties to task properties? I don't think it would. Let's assume a `FileCollection` is-a `Provider<Set<File>>` and the following code was used:

```
class MyExtension {
    Boolean enabled
    FileCollection outputFiles

    MyExtension(Project project) {
        outputFiles = project.files()
    }
}

project.tasks.create('myTask', MyTask) {
    enabled = project.provider { extension.enabled }
    outputFiles = extension.outputFiles
}
```

When the plugin is applied that creates `myTask` then the mapped value for `outputFiles` would be an empty `FileCollection`. However, when the consuming build script also sets the value of `outputFiles` with the extension then it wouldn't propagate down to the task anymore. The value of `outFiles` for the task would still be the empty `FileCollection`.

```
pluginConfig {
    enabled = true
    outputFiles = files('custom.txt')
}
```

Am I missing something or how do you think the use case can be fulfilled? Using a `project.provider { extension.outputFiles }` works because the `Callable` is going to be evaluated when the task is actually executed.